syntax = "proto3";
import "google/protobuf/wrappers.proto";
package kvraft;

message Operation{
    enum OpName{
        GET = 0;
        PUT = 1;
        DEL = 2;
    }
    int32 index = 1;
    OpName op = 2;
    string key = 3;
    string value = 4;
}

enum KVResult{
    OK       = 0;
	ErrNoKey = 1;
	ErrTimeout = 2;
	ErrWrongLeader = 3;
}

message KVArgs{
    Operation command = 1;
    int32 clientid = 2;
    int32 seq = 3;
}

message KVReply{
    KVResult res = 1;
    string value = 2;
}

message LogEntry{
    Operation command = 1;
    int32 term = 2;
}

message RequestVoteArgs{
    int32 term = 1;
    int32 candidateid = 2;
    int32 lastlogindex = 3;
    int32 lastlogterm = 4; 
}

message RequestVoteReply{
    int32 term = 1;
    bool votegranted = 2;
}

message AppendEntriesArgs{
    int32 term = 1;
    int32 leaderid = 2;
    int32 prevlogindex = 3;
    int32 prevlogterm = 4;
    int32 leadercommit = 5;
    repeated LogEntry entries = 6;
}

message AppendEntriesReply{
    int32 term = 1;
    bool success = 2;
    int32 replicatedindex = 3;
}

service KVRaft{
    rpc RequestVote(RequestVoteArgs) returns (RequestVoteReply);
    rpc AppendEntries(AppendEntriesArgs) returns (AppendEntriesReply);
    rpc Command(KVArgs) returns (KVReply);
}